File: python3/Fast and Slow Pointers/linked_list_loop.py
from ds import ListNode

"""
Definition of ListNode:
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


def linked_list_loop(head: ListNode) -> bool:
    slow = fast = head
    # Check both 'fast' and 'fast.next' to avoid null pointer
    # exceptions when we perform 'fast.next' and 'fast.next.next'.
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if fast == slow:
            return True
    return False

File: kotlin/Fast and Slow Pointers/LinkedListLoop.kt
/*
    Definition of ListNode:
    data class ListNode(var value: Int, var next: ListNode? = null)
*/

fun linkedListLoop(head: ListNode?): Boolean {
    var slow = head
    var fast = head
    // Check both 'fast' and 'fast.next' to avoid null pointer
    // exceptions when we perform 'fast.next' and 'fast.next.next'.
    while (fast != null && fast.next != null) {
        slow = slow!!.next
        fast = fast.next!!.next
        if (fast == slow) {
            return true
        }
    }
    return false
}